File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

using DevExpress.XtraGrid.Views.Grid;

namespace LoadExternalImages {
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form {
      private DevExpress.XtraGrid.GridControl gridControl1;
      private DevExpress.XtraGrid.Views.Grid.GridView gridView1;
      private System.Windows.Forms.Label label1;
      private DevExpress.XtraGrid.Columns.GridColumn columnID;
      private DevExpress.XtraGrid.Columns.GridColumn columnName;
      private DevExpress.XtraGrid.Columns.GridColumn columnColor;
      private DevExpress.XtraGrid.Columns.GridColumn columnImage;
      private DevExpress.XtraEditors.Repository.RepositoryItemPictureEdit repositoryItemPictureEdit1;
      private DevExpress.XtraEditors.Repository.RepositoryItemComboBox repositoryItemComboBox1;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
         InitData();
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.gridControl1 = new DevExpress.XtraGrid.GridControl();
            this.gridView1 = new DevExpress.XtraGrid.Views.Grid.GridView();
            this.columnID = new DevExpress.XtraGrid.Columns.GridColumn();
            this.columnName = new DevExpress.XtraGrid.Columns.GridColumn();
            this.columnColor = new DevExpress.XtraGrid.Columns.GridColumn();
            this.repositoryItemComboBox1 = new DevExpress.XtraEditors.Repository.RepositoryItemComboBox();
            this.columnImage = new DevExpress.XtraGrid.Columns.GridColumn();
            this.repositoryItemPictureEdit1 = new DevExpress.XtraEditors.Repository.RepositoryItemPictureEdit();
            this.label1 = new System.Windows.Forms.Label();
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.repositoryItemComboBox1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.repositoryItemPictureEdit1)).BeginInit();
            this.SuspendLayout();
            // 
            // gridControl1
            // 
            this.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill;
            this.gridControl1.EmbeddedNavigator.Name = "";
            this.gridControl1.Location = new System.Drawing.Point(0, 23);
            this.gridControl1.LookAndFeel.SkinName = "The Asphalt World";
            this.gridControl1.LookAndFeel.UseDefaultLookAndFeel = false;
            this.gridControl1.MainView = this.gridView1;
            this.gridControl1.Name = "gridControl1";
            this.gridControl1.RepositoryItems.AddRange(new DevExpress.XtraEditors.Repository.RepositoryItem[] {
            this.repositoryItemPictureEdit1,
            this.repositoryItemComboBox1});
            this.gridControl1.Size = new System.Drawing.Size(527, 286);
            this.gridControl1.TabIndex = 0;
            this.gridControl1.ViewCollection.AddRange(new DevExpress.XtraGrid.Views.Base.BaseView[] {
            this.gridView1});
            // 
            // gridView1
            // 
            this.gridView1.Columns.AddRange(new DevExpress.XtraGrid.Columns.GridColumn[] {
            this.columnID,
            this.columnName,
            this.columnColor,
            this.columnImage});
            this.gridView1.GridControl = this.gridControl1;
            this.gridView1.Name = "gridView1";
            this.gridView1.OptionsView.ShowGroupPanel = false;
            this.gridView1.RowHeight = 33;
            this.gridView1.CustomUnboundColumnData += new DevExpress.XtraGrid.Views.Base.CustomColumnDataEventHandler(this.gridView1_CustomUnboundColumnData);
            // 
            // columnID
            // 
            this.columnID.Caption = "ID";
            this.columnID.FieldName = "ID";
            this.columnID.Name = "columnID";
            this.columnID.OptionsColumn.ReadOnly = true;
            this.columnID.Visible = true;
            this.columnID.VisibleIndex = 0;
            this.columnID.Width = 81;
            // 
            // columnName
            // 
            this.columnName.Caption = "Name";
            this.columnName.FieldName = "Name";
            this.columnName.Name = "columnName";
            this.columnName.Visible = true;
            this.columnName.VisibleIndex = 1;
            this.columnName.Width = 132;
            // 
            // columnColor
            // 
            this.columnColor.Caption = "Color";
            this.columnColor.ColumnEdit = this.repositoryItemComboBox1;
            this.columnColor.FieldName = "Color";
            this.columnColor.Name = "columnColor";
            this.columnColor.Visible = true;
            this.columnColor.VisibleIndex = 2;
            this.columnColor.Width = 152;
            // 
            // repositoryItemComboBox1
            // 
            this.repositoryItemComboBox1.AutoHeight = false;
            this.repositoryItemComboBox1.Buttons.AddRange(new DevExpress.XtraEditors.Controls.EditorButton[] {
            new DevExpress.XtraEditors.Controls.EditorButton(DevExpress.XtraEditors.Controls.ButtonPredefines.Combo)});
            this.repositoryItemComboBox1.Items.AddRange(new object[] {
            "Black",
            "Blue",
            "Green",
            "Magenta",
            "Orange",
            "Pink",
            "Red",
            "White",
            "Yellow"});
            this.repositoryItemComboBox1.Name = "repositoryItemComboBox1";
            // 
            // columnImage
            // 
            this.columnImage.Caption = "Image (unbound)";
            this.columnImage.ColumnEdit = this.repositoryItemPictureEdit1;
            this.columnImage.FieldName = "Image";
            this.columnImage.Name = "columnImage";
            this.columnImage.OptionsColumn.AllowEdit = false;
            this.columnImage.UnboundType = DevExpress.Data.UnboundColumnType.Object;
            this.columnImage.Visible = true;
            this.columnImage.VisibleIndex = 3;
            this.columnImage.Width = 118;
            // 
            // repositoryItemPictureEdit1
            // 
            this.repositoryItemPictureEdit1.Name = "repositoryItemPictureEdit1";
            // 
            // label1
            // 
            this.label1.Dock = System.Windows.Forms.DockStyle.Top;
            this.label1.Font = new System.Drawing.Font("Microsoft Sans Serif", 9F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ((byte)(204)));
            this.label1.Location = new System.Drawing.Point(0, 0);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(527, 23);
            this.label1.TabIndex = 1;
            this.label1.Text = "What\'s your favorite color?";
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(527, 309);
            this.Controls.Add(this.gridControl1);
            this.Controls.Add(this.label1);
            this.Name = "Form1";
            this.Text = "Form1";
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.repositoryItemComboBox1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.repositoryItemPictureEdit1)).EndInit();
            this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		private void InitData() {
			string DBFileName = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, "db.mdb");
			if(DBFileName != "") {
				DataSet ds = new DataSet();
				string con = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source=" + DBFileName;
				System.Data.OleDb.OleDbDataAdapter oleDbDataAdapter = new System.Data.OleDb.OleDbDataAdapter("SELECT * FROM PeopleColors", con);
				oleDbDataAdapter.Fill(ds, "PeopleColors");
				gridControl1.DataSource = ds.Tables["PeopleColors"];
			}
		}

		string ImageDir = @"Images\";
		Hashtable Images = new Hashtable();

		string GetFileName(string color) {         
			if(color == null || color == string.Empty)
				return string.Empty;				
			return color + ".jpg";			
		}

		private void gridView1_CustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e) {
			if(e.Column.FieldName == "Image" && e.IsGetData) {
				GridView view = sender as GridView;

				string colorName = (string)view.GetRowCellValue(e.RowHandle, "Color");
				string fileName = GetFileName(colorName).ToLower();
				if(!Images.ContainsKey(fileName)) {
					Image img = null;
					try {
						string filePath = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, ImageDir + fileName, false);
						img = Image.FromFile(filePath);
					}
					catch {
					}
					Images.Add(fileName, img);
				}
				e.Value = Images[fileName];
			}
		}
	}
}

File: VB/Form1.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Data.OleDb

Namespace LoadExternalImages

    ''' <summary>
    ''' Summary description for Form1.
    ''' </summary>
    Public Class Form1
        Inherits Form

        Private gridControl1 As DevExpress.XtraGrid.GridControl

        Private gridView1 As GridView

        Private label1 As Label

        Private columnID As DevExpress.XtraGrid.Columns.GridColumn

        Private columnName As DevExpress.XtraGrid.Columns.GridColumn

        Private columnColor As DevExpress.XtraGrid.Columns.GridColumn

        Private columnImage As DevExpress.XtraGrid.Columns.GridColumn

        Private repositoryItemPictureEdit1 As DevExpress.XtraEditors.Repository.RepositoryItemPictureEdit

        Private repositoryItemComboBox1 As DevExpress.XtraEditors.Repository.RepositoryItemComboBox

        ''' <summary>
        ''' Required designer variable.
        ''' </summary>
        Private components As System.ComponentModel.Container = Nothing

        Public Sub New()
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()
            '
            ' TODO: Add any constructor code after InitializeComponent call
            '
            InitData()
        End Sub

        ''' <summary>
        ''' Clean up any resources being used.
        ''' </summary>
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If components IsNot Nothing Then
                    components.Dispose()
                End If
            End If

            MyBase.Dispose(disposing)
        End Sub

'#Region "Windows Form Designer generated code"
        ''' <summary>
        ''' Required method for Designer support - do not modify
        ''' the contents of this method with the code editor.
        ''' </summary>
        Private Sub InitializeComponent()
            gridControl1 = New DevExpress.XtraGrid.GridControl()
            gridView1 = New GridView()
            columnID = New DevExpress.XtraGrid.Columns.GridColumn()
            columnName = New DevExpress.XtraGrid.Columns.GridColumn()
            columnColor = New DevExpress.XtraGrid.Columns.GridColumn()
            repositoryItemComboBox1 = New DevExpress.XtraEditors.Repository.RepositoryItemComboBox()
            columnImage = New DevExpress.XtraGrid.Columns.GridColumn()
            repositoryItemPictureEdit1 = New DevExpress.XtraEditors.Repository.RepositoryItemPictureEdit()
            label1 = New Label()
            CType(gridControl1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(gridView1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(repositoryItemComboBox1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(repositoryItemPictureEdit1, System.ComponentModel.ISupportInitialize).BeginInit()
            Me.SuspendLayout()
            ' 
            ' gridControl1
            ' 
            gridControl1.Dock = DockStyle.Fill
            gridControl1.EmbeddedNavigator.Name = ""
            gridControl1.Location = New System.Drawing.Point(0, 23)
            gridControl1.LookAndFeel.SkinName = "The Asphalt World"
            gridControl1.LookAndFeel.UseDefaultLookAndFeel = False
            gridControl1.MainView = gridView1
            gridControl1.Name = "gridControl1"
            gridControl1.RepositoryItems.AddRange(New DevExpress.XtraEditors.Repository.RepositoryItem() {repositoryItemPictureEdit1, repositoryItemComboBox1})
            gridControl1.Size = New System.Drawing.Size(527, 286)
            gridControl1.TabIndex = 0
            gridControl1.ViewCollection.AddRange(New DevExpress.XtraGrid.Views.Base.BaseView() {gridView1})
            ' 
            ' gridView1
            ' 
            gridView1.Columns.AddRange(New DevExpress.XtraGrid.Columns.GridColumn() {columnID, columnName, columnColor, columnImage})
            gridView1.GridControl = gridControl1
            gridView1.Name = "gridView1"
            gridView1.OptionsView.ShowGroupPanel = False
            gridView1.RowHeight = 33
            AddHandler gridView1.CustomUnboundColumnData, New DevExpress.XtraGrid.Views.Base.CustomColumnDataEventHandler(AddressOf gridView1_CustomUnboundColumnData)
            ' 
            ' columnID
            ' 
            columnID.Caption = "ID"
            columnID.FieldName = "ID"
            columnID.Name = "columnID"
            columnID.OptionsColumn.ReadOnly = True
            columnID.Visible = True
            columnID.VisibleIndex = 0
            columnID.Width = 81
            ' 
            ' columnName
            ' 
            columnName.Caption = "Name"
            columnName.FieldName = "Name"
            columnName.Name = "columnName"
            columnName.Visible = True
            columnName.VisibleIndex = 1
            columnName.Width = 132
            ' 
            ' columnColor
            ' 
            columnColor.Caption = "Color"
            columnColor.ColumnEdit = repositoryItemComboBox1
            columnColor.FieldName = "Color"
            columnColor.Name = "columnColor"
            columnColor.Visible = True
            columnColor.VisibleIndex = 2
            columnColor.Width = 152
            ' 
            ' repositoryItemComboBox1
            ' 
            repositoryItemComboBox1.AutoHeight = False
            repositoryItemComboBox1.Buttons.AddRange(New DevExpress.XtraEditors.Controls.EditorButton() {New DevExpress.XtraEditors.Controls.EditorButton(DevExpress.XtraEditors.Controls.ButtonPredefines.Combo)})
            repositoryItemComboBox1.Items.AddRange(New Object() {"Black", "Blue", "Green", "Magenta", "Orange", "Pink", "Red", "White", "Yellow"})
            repositoryItemComboBox1.Name = "repositoryItemComboBox1"
            ' 
            ' columnImage
            ' 
            columnImage.Caption = "Image (unbound)"
            columnImage.ColumnEdit = repositoryItemPictureEdit1
            columnImage.FieldName = "Image"
            columnImage.Name = "columnImage"
            columnImage.OptionsColumn.AllowEdit = False
            columnImage.UnboundType = DevExpress.Data.UnboundColumnType.Object
            columnImage.Visible = True
            columnImage.VisibleIndex = 3
            columnImage.Width = 118
            ' 
            ' repositoryItemPictureEdit1
            ' 
            repositoryItemPictureEdit1.Name = "repositoryItemPictureEdit1"
            ' 
            ' label1
            ' 
            label1.Dock = DockStyle.Top
            label1.Font = New System.Drawing.Font("Microsoft Sans Serif", 9F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, (CByte(204)))
            label1.Location = New System.Drawing.Point(0, 0)
            label1.Name = "label1"
            label1.Size = New System.Drawing.Size(527, 23)
            label1.TabIndex = 1
            label1.Text = "What's your favorite color?"
            ' 
            ' Form1
            ' 
            AutoScaleBaseSize = New System.Drawing.Size(5, 13)
            ClientSize = New System.Drawing.Size(527, 309)
            Me.Controls.Add(gridControl1)
            Me.Controls.Add(label1)
            Name = "Form1"
            Text = "Form1"
            CType(gridControl1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(gridView1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(repositoryItemComboBox1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(repositoryItemPictureEdit1, System.ComponentModel.ISupportInitialize).EndInit()
            Me.ResumeLayout(False)
        End Sub

'#End Region
        ''' <summary>
        ''' The main entry point for the application.
        ''' </summary>
        <STAThread>
        Shared Sub Main()
            Call Application.Run(New Form1())
        End Sub

        Private Sub InitData()
            Dim DBFileName As String = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, "db.mdb")
            If Not Equals(DBFileName, "") Then
                Dim ds As DataSet = New DataSet()
                Dim con As String = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source=" & DBFileName
                Using oleDbDataAdapter = New OleDbDataAdapter("SELECT * FROM PeopleColors", con)
                    oleDbDataAdapter.Fill(ds, "PeopleColors")
                End Using

                gridControl1.DataSource = ds.Tables("PeopleColors")
            End If
        End Sub

        Private ImageDir As String = "Images\"

        Private Images As Hashtable = New Hashtable()

        Private Function GetFileName(ByVal color As String) As String
            If Equals(color, Nothing) OrElse Equals(color, String.Empty) Then Return String.Empty
            Return color & ".jpg"
        End Function

        Private Sub gridView1_CustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs)
            If Equals(e.Column.FieldName, "Image") AndAlso e.IsGetData Then
                Dim view As GridView = TryCast(sender, GridView)
                Dim colorName As String = CStr(CType(e.Row, DataRowView)("Color"))
                Dim fileName As String = GetFileName(colorName).ToLower()
                If Not Images.ContainsKey(fileName) Then
                    Dim img As Image = Nothing
                    Try
                        Dim filePath As String = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, ImageDir & fileName, False)
                        img = Image.FromFile(filePath)
                    Catch
                    End Try

                    Images.Add(fileName, img)
                End If

                e.Value = Images(fileName)
            End If
        End Sub
    End Class
End Namespace
